### ISSUE TITLE
Follow XDG for global config files

### ISSUE THREAD

**Issue by samoht** [ocaml-ppx/ocamlformat#361]
[XDG](https://standards.freedesktop.org/basedir-spec/basedir-spec-latest.html) is becoming the standard to store configuration files (and it's followed by dune :p). The specs says the global config file should be something like: `./config/<toolname>/file`

**Comment by samoht** [comment]
Basically XDG states that the proliferation of global dot files are annoying, and should all be moved to a `~/.config/` subdir instead. So yes, with XDG `~/.ocamlformat` will have to move to `~/.config/ocamlformat/config` or ~/.config/ocaml/ocamlformat.config` or something like that.

**Comment by jberdine** [comment]
I agree with the basic motivation behind XDG. The part I'm unsure of in the context of ocamlformat is that its configuration is intimately tied source code much more than individual users, at least in the use cases I know of. In particular, the ocamlformat configuration needs to be kept in sync with the source code via source control. I was actually thinking before about how to ensure that no configuration is used that is outside the source control repo.

**Comment by hhugo** [comment]
I kind of agree with @jberdine. Not reading configuration outside the "project" seems right. I even wonder if ocamlformat should simply be disabled if no `.ocamlformat` is found in the project.

**Comment by samoht** [comment]
>  I even wonder if ocamlformat should simply be disabled if no .ocamlformat is found in the project.

I think that's too strict. Most project will just use the "default" formating style, which IMHO needs to be a bit redesigned. The goal of ocamlformat should be to allow the different styles of the community to converge to an "acceptable" default, not to have every project using a different style.

**Comment by hhugo** [comment]
Note, that project can still use whatever default by having an empty `.ocamlformat` in their project.
I agree with the goal of convergence, but in practice, I do not use ocamlformat at home because all code bases I touch (and most often don't own) have different style and I don't want Emacs to reformat entire files on save.

**Comment by jberdine** [comment]
(Note that unless it has changed, adding an empty `.ocamlformat` file just passes the config from the parent through, rather than sets options to defaults.)

I agree that the eventual goal is for the community to converge on a (few) common styles, maybe even one, with a "default" that people who are not overly picky about syntax can get along with. I think the intent is for the config files to move to just specifying one of the umbrella styles, perhaps with a few exceptions. But I don't understand how this is connected to the question of whether to read formatting configuration from within the source tree. It seems that, as something that if changed, changes the source, the formatter config needs to be part of the source tree.

Optimally there would be tools that allow each person to smoothly interact with code in their preferred style while interacting with the community (that is, source control) in some probably different style. If we had such tools, then it would make sense to override the source tree's config. But presuming that source control still stores text, it will be in some style that needs to be shared by all contributions, so the source tree config will still be needed.

Maybe I'm missing the point you're making?

**Comment by jberdine** [comment]
Ok, I see. I agree that it should be fully-functional without a config file, just use the default. And I think I have the "monorepo blinders" on where I assume "a shared global config directory" is in the repo, hence the thought that config shouldn't come from outside the repo. But there are other approaches too.

In the situation where there is some config in e.g. ~/.config/ocamlformat/config and some .ocamlformat in the repo root, my expectation is that if the repo's .ocamlformat specifies `style = something-non-standard`, then it sets _all_ the options, effectively overriding the extra-repo config. Is that compatible with what you have in mind?

**Comment by samoht** [comment]
> In the situation where there is some config in e.g. ~/.config/ocamlformat/config and some .ocamlformat in the repo root, my expectation is that if the repo's .ocamlformat specifies style = something-non-standard, then it sets all the options, effectively overriding the extra-repo config. Is that compatible with what you have in mind?

yes, that makes sense.
